### PR TITLE
quartus: init at 19.1.0.670

### PR DESCRIPTION
**Pull request by kwohlfahrt**
###### Motivation for this change

Add Quartus Prime software for programming Intel FPGAs. There are unfortunately no open-source alternatives.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

This is a new package - should i add myself as a maintainer, even though I don't have commit access to nixpkgs?

### CODE REVIEW

**Comment by kwohlfahrt** [comment]
This also adds a utility that silently eats chmod syscalls that set SUID/SGID bits. The installer insists on setting this on directories, which nix then causes to fail.

Nix could in theory allow this, I think the action is safe on directories but I'm not sure it is possible to intercept the syscall, check arguments and then proceed in a way that does not produce a race condition (where a directory is replaced with a file after the check but before the operation takes place).

File: pkgs/os-specific/linux/unstick/default.nix
Comment (doronbehar) [inline pkgs/os-specific/linux/unstick/default.nix]: It's wonderful you've created this software especially for Nix. Your contribution is appreciated! Please quote the URL and remove the `.git` suffix.

**Review by doronbehar — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
> This is a new package - should i add myself as a maintainer, even though I don't have commit access to nixpkgs?

Yes, it's very common to maintain while not having commit access. Please add yourself as a maintainer to the maintainers file and put that addition in a separate commit.

Other then that and my 1 comment, LGTM. Just out of curiosity, is this your 1st contribution to Nixpkgs? It seems you have used advanced mechanisms to package this, similar to WolframAlpha's package.. Well done!

File: pkgs/os-specific/linux/unstick/default.nix
Comment (kwohlfahrt) [reply pkgs/os-specific/linux/unstick/default.nix]: The `.git` is definitely out of place, fixed.

The majority of URLs on the `homepage` attribute across nixpkgs are unquoted (10k unquoted, 2k quoted according to `git grep`), is there any style guide for things like this?

**Comment by kwohlfahrt** [comment]
Thanks for the review! This isn't my first contribution to nixpkgs, but it's definitely my trickiest so far, open-source is generally much easier.

I've added myself to the maintainers.

File: pkgs/os-specific/linux/unstick/default.nix
Comment (doronbehar) [reply pkgs/os-specific/linux/unstick/default.nix]: Yes. See https://github.com/NixOS/rfcs/pull/45

Comment (kwohlfahrt) [reply pkgs/os-specific/linux/unstick/default.nix]: Cool, fixed.

**Comment by kwohlfahrt** [comment]
Is it possible to have more than one executable in a FHS user environment? The `quartus` program really comes with several binaries that should be runnable from the command line.

**Comment by doronbehar** [comment]
Good question. I'm a fan myself of the command line tools that come with Quartus and I used them back when I was on Arch and I developed stuff for FPGA.

It's a good question to ask on the forum. The main issue is that there's the `runScript` argument which should be mandatory for this.

1 workaround I have in mind while thinking about it, is this:

Create a wrapper script that will replace the current `runScript`. Make this wrapper check it's `$0` and accordingly launch the right executable in the FHS env. Afterwards, create a wrapper _package_  that will create symlinks to that FHS' wrapper script and use different names in `$out/bin/`. This way, the symlinks will emulate a different `$0` and thus operate a different executables in the FHS environment.

I hope you can imagine what I'm describing. From AFAIK this is a common practice for example with rustup - there's only 1 executable and it checks it's `$0` in order to operate as `rustc` or `cargo` or whatever. The same is done also with bash and `sh` - if it's `$0` is `sh` it'll operate in a `sh` compatible mode.

**Comment by kwohlfahrt** [comment]
OK, I'll try and put something together. I'd like it to work with `.desktop` files so I can use it nicely from a menu.

Secondly, if you have much experience with Quartus - the "Platform Designer" tool does not work at least when launched from the main UI. Unfortunately, there are no GUI or console error logs, I just get the Platform Designer splash screen and then nothing. Do you know either which binary this is, so I can try running it directly, or if there are any more detailed logs kept anywhere?

I wouldn't consider this a blocker but it would be nice to fix.

**Comment by doronbehar** [comment]
I'm glad you worked this out @kwohlfahrt! as for the other command line interface executables, could you perhaps just verify, as a start, that the command line tools [listed here](https://www.intel.com/content/dam/www/programmable/us/en/pdfs/literature/manual/tclscriptrefmnl.pdf) are runnable from inside the FHS env? You can test it by temporarily replacing the `runScript` with `bash` and then see if you can launch them from that shell..

If they all work fine, I think we can work out together a way to make all of these executables available in the user's `$PATH` with a wrapper package and a wrapper script, as I tried to explain in my [previous comment](https://github.com/NixOS/nixpkgs/pull/75561#issuecomment-565758484).

**Comment by kwohlfahrt** [comment]
`quartus/bin/jtagconfig` works, `quartus/bin/quartus` obviously does, `quartus/sopc_builder/bin/qsys-edit` is OK. Any others I should check?

`modelsim_ase/bin/vsim` exits with `Read failure in vlm process`. I assume I also have to set-up some licensing things to run this directly, since it works when launched from the main UI.

**Comment by doronbehar** [comment]
Good job @kwohlfahrt. There are many more executables it would be nice to check (e.g `quartus_asm`, `quartus_cdb`...) - please see the list in this [PDF's TOC](https://www.intel.com/content/dam/www/programmable/us/en/pdfs/literature/manual/tclscriptrefmnl.pdf).

As for the modelsim error, see this suggested solution:

https://bbs.archlinux.org/viewtopic.php?pid=1708875#p1708875

**Comment by Mic92** [comment]
Is this ready for review?

**Comment by kwohlfahrt** [comment]
@Mic92 I have been using this package via the main GUI for a few months, it seems to be working OK. However, I have not tested many of the individual executables suggested above. Hoping to do so soon now I'm back from the holiday, so I don't mind deferring this until I have, but I also wouldn't object to merging this now and fixing any issues found later.

**Comment by doronbehar** [comment]
@kwohlfahrt From my experience with the PRs culture in Nixpkgs, I think it'll be worth marking this as WIP untill it'll be fully ready.

**Comment by kwohlfahrt** [comment]
The following work: `../sopc_builder/bin/qsys-generate`, `./quartus_map`, `quartus_fit`, `quartus_asm`, `quartus_sta`, `quartus_eda`. Some others give me errors that they are not available with my license, or that I am missing command-line arguments but they do run.

`vsim` still does not work - however Ubuntu 18.04 is supported which has `freetype-2.8.1`, so it is not *exactly* the same issue. I will try and dig into what exactly the missing factor is here.

**Comment by Mic92** [comment]
I don't have an Intel account to download/test this but would merge this pull request if somebody else was able to launch the GUI i.e. @doronbehar

File: pkgs/applications/editors/quartus-prime/default.nix
Comment (doronbehar) [inline pkgs/applications/editors/quartus-prime/default.nix]: BTW I'd suggest adding here that it can also be done via `nix-store --add <file>`

**Comment by doronbehar** [comment]
I thought up until now I'll need a paid license in order to test this but it seems it's supposed to be possible to download the what's called ["standard" edition](https://fpgasoftware.intel.com/19.1/?edition=standard) of the `QuartusLiteSetup-${version}-linux.run` files and that's what the derivation is targeted at - the only requirement is to sign up for a free account. Here are my findings:

I downloaded these `.run` files and added them with `nix-store --add` (see my review about this) and I think because my `sha256sum` were different I still got the message that I need to put the ModelSim .run file in my `/nix/store/`. The shas seem to be different for the QuartusLiteSetup `.run` file as well:

```
$ ls -l /nix/store/4yihv30am9640ixijz0ip7x4msnlwshx-QuartusSetup-19.1.0.670-linux.run /nix/store/gy4maz398wv3d8hrf3j95zrcddrrmlig-QuartusLiteSetup-19.1.0.670-linux.run.drv
-r--r--r-- 1 root root 2701700693 Jan  1  1970 /nix/store/4yihv30am9640ixijz0ip7x4msnlwshx-QuartusSetup-19.1.0.670-linux.run
-r--r--r-- 1 root root       1486 Jan  1  1970 /nix/store/gy4maz398wv3d8hrf3j95zrcddrrmlig-QuartusLiteSetup-19.1.0.670-linux.run.drv
$ ls -l /nix/store/y75qnagk59qfhpc5sla83iv7g4jnxdb9-ModelSimSetup-19.1.0.670-linux.run /nix/store/hj9gvpmzjszrcq5yndq6jjs8n7jrcdkl-ModelSimSetup-19.1.0.670-linux.run.drv /nix/store/0bi68pph1sn7c1dgfg4d4dyvfglds1lv-ModelSimSetup-19.1.0.670-linux.run.lock
-rw------- 1 root root          0 Jan 10 17:47 /nix/store/0bi68pph1sn7c1dgfg4d4dyvfglds1lv-ModelSimSetup-19.1.0.670-linux.run.lock
-r--r--r-- 1 root root       1477 Jan  1  1970 /nix/store/hj9gvpmzjszrcq5yndq6jjs8n7jrcdkl-ModelSimSetup-19.1.0.670-linux.run.drv
-r--r--r-- 1 root root 1047232539 Jan  1  1970 /nix/store/y75qnagk59qfhpc5sla83iv7g4jnxdb9-ModelSimSetup-19.1.0.670-linux.run

```

**Comment by Mic92** [comment]
@kwohlfahrt can you confirm the different checksums?

**Comment by kwohlfahrt** [comment]
Sorry, missed the previous comment. I get the following checksums for the `ModelSimSetup` file. The `md5sum` matches that published on the download website:

```
> md5sum ModelSimSetup-19.1.0.670-linux.run 
142526799b9f80e1fe1d11743724a668  ModelSimSetup-19.1.0.670-linux.run
> sha256sum ModelSimSetup-19.1.0.670-linux.run
d9f841c12a0d7360222068610dca23bff6ac90e9f48aaa2c429b321952763b48  ModelSimSetup-19.1.0.670-linux.run
```

`nix-prefetch-url` produces the checksum added in the PR. `nix-store --add` however comes up with a different store path, I'm not sure what's up with that.

```
> nix-prefetch-url file://(pwd)/ModelSimSetup-19.1.0.670-linux.run
warning: dumping very large path (> 256 MiB); this may run out of memory
[998.7 MiB DL]
path is '/nix/store/0bi68pph1sn7c1dgfg4d4dyvfglds1lv-ModelSimSetup-19.1.0.670-linux.run'
0j1vfr91jclv88nam2plx68arxmz4g50sqb840i60wqd5b0l3y6r
> nix-store --add ./ModelSimSetup-19.1.0.670-linux.run 
warning: dumping very large path (> 256 MiB); this may run out of memory
/nix/store/b6cc9qpmiq8h441984p555601s3w4p2v-ModelSimSetup-19.1.0.670-linux.run
```

I notice the hash part of the path from `nix-prefetch-url` matches the `.lock` file from @doronbehar above.

Additionally, `vsim` works if I apply the following overlay: `self: super: { libpng = super.libpng12; }`. Is there a way to apply this to the `multiPkgs` of `buildFHSUserEnv`?

File: pkgs/applications/editors/quartus-prime/default.nix
Comment (kwohlfahrt) [reply pkgs/applications/editors/quartus-prime/default.nix]: Sure, is there any reason for preferring one over the other? Going to hold off on making any changes until we get the hash issue sorted out though.

File: pkgs/os-specific/linux/unstick/default.nix
Comment (kwohlfahrt) [reply pkgs/os-specific/linux/unstick/default.nix]: Fixed.

**Comment by kwohlfahrt** [comment]
Given the above requirement regarding libpng, I'm tempted to split ModelSim into its own package. @doronbehar, do you have any idea whether this should present difficulties? As far as I know Quartus just calls the ModelSim executables so as long as we preserve those it should be OK...

Rebased onto the wrong branch, sorry for stressing out poor ofborg.

**Comment by doronbehar** [comment]
> I'm tempted to split ModelSim into its own package.

It's an interesting idea but I think it doesn't worth the effort and the risk of whatever issues users may encounter in the far future..

> Additionally, vsim works if I apply the following overlay: self: super: { libpng = super.libpng12; }. Is there a way to apply this to the multiPkgs of buildFHSUserEnv?

Why not package the whole derivation with `libpng12` instead of `libpng`?

> nix-prefetch-url produces the checksum added in the PR. nix-store --add however comes up with a different store path, I'm not sure what's up with that.

That's definitely peculiar and it's great you've summoned more reviewers who may be able to shed some light on this. Never the less, as long as we may end up stuck with this inconsistency, I think it's less intuitive to go along with `nix-prefetch-url` as the files are usually downloaded with the browser.. I wouldn't mind much if we'll go along with either of these options but perhaps at least specify a little bit better how to use either command. For example, I had no clue that `nix-prefetch-url` could be used with a `file://` schema. You see where I'm getting at @kwohlfahrt ?

**Comment by Mic92** [comment]
In that case the error message should explicitly list on how to add those file to the nix store.

**Comment by Mic92** [comment]
@doronbehar can you re-check?

File: pkgs/applications/editors/quartus-prime/default.nix
Comment (kwohlfahrt) [reply pkgs/applications/editors/quartus-prime/default.nix]: I've removed the custom message. The default message now suggests `nix-store --add-fixed sha256` or `nix-prefetch-url --type sha256 file://path/to`. C

**Comment by kwohlfahrt** [comment]
I got rid of the custom message. Note that `nix-store --add` produces a different store path compared to `nix-store --add-fixed sha256` and `nix-prefetch-url`. Could you try again using one of the latter two methods @doronbehar?

Regarding `vsim`/`libpng12`: Adding `libpng12` to `multipkgs` is insufficient, I still get the error below.

```
Initialization problem, exiting.

    while executing
"InitializeINIFile quietly"
    invoked from within
"ncFyP12 -+"
    (file "/mtitcl/vsim/vsim" line 1)
** Fatal: Read failure in vlm process (0,0)
```

I assume this occurs if `libpng` is linked anywhere instead of `libpng12` - I'm not sure how these versions are resolved at runtime. So far, the only fix I've found is the overlay approach, so all libraries that depend on `libpng` are rebuilt with `libpng12`. For example, `freetype` depends on `libpng`, so that must be overridden. But then `fontconfig` depends on `freetype`, so I need to override `fontconfig`'s `freetype` dependency with the one we just built. I imagine there are a fair few `libpng` dependencies buried somewhere in the X11 requirements as well, so I don't think it is practical to do this one package at a time.

This is also my reasoning for splitting out the ModelSim package - these overridden won't be available from the Nix cache, and I would prefer to use the up-to-date libraries where they don't cause issues.

PS - I didn't summon the extra reviewers, that happened automatically.

**Comment by kwohlfahrt** [comment]
I think I got `vsim` working! The overrides are a bit ugly the way they are now though, can this be applied to all of `pkgs` cleanly somehow?

**Comment by doronbehar** [comment]
@kwohlfahrt Well done for finding out about `nix-store --add-fixed` and changing the message. I've readded the files with:

```
~/downloads $ nix-store --add-fixed sha256 ./QuartusSetup-19.1.0.670-linux.run ./ModelSimSetup-19.1.0.670-linux.run
```

When building the derivation, I got no error as for the ModelSim setup `.run` file but only for Quartus. It's seems as if it's because the file names are different - mine is `./QuartusSetup-19.1.0.670-linux.run` and the expected filename is `QuartusLiteSetup-19.1.0.670-linux.run`. After renaming the file and running `nix-store --add-fixed` again, I tried building but the message came up again. Could you recheck please with Quartus' download on your end what is the filename that's used when you download the file?

Here are the files I have in my `/nix/store` right now:

```
$ ls -l /nix/store/*Quartus*
-r--r--r-- 1 root root 2701700693 Jan  1  1970 /nix/store/4yihv30am9640ixijz0ip7x4msnlwshx-QuartusSetup-19.1.0.670-linux.run
-rw------- 1 root root          0 Jan 13 19:06 /nix/store/9m7nc46lcrjhr74lag3h2j8wlxbibk6g-QuartusLiteSetup-19.1.0.670-linux.run.lock
-r--r--r-- 1 root root       1486 Jan  1  1970 /nix/store/gy4maz398wv3d8hrf3j95zrcddrrmlig-QuartusLiteSetup-19.1.0.670-linux.run.drv
-r--r--r-- 1 root root 2701700693 Jan  1  1970 /nix/store/pnn5bvcz7q6cp5zqf77b66csw0h4qn48-QuartusSetup-19.1.0.670-linux.run
-r--r--r-- 1 root root 2701700693 Jan  1  1970 /nix/store/svq7aag9lc5mmbp0rr00qfqwk38hmf3k-QuartusLiteSetup-19.1.0.670-linux.run
-r--r--r-- 1 root root       1486 Jan  1  1970 /nix/store/waznhb972qp8nfj2h70zqg61b7z9ps8a-QuartusLiteSetup-19.1.0.670-linux.run.drv
$ ls -l /nix/store/*ModelSim*
-r--r--r-- 1 root root 1047232539 Jan  1  1970 /nix/store/0bi68pph1sn7c1dgfg4d4dyvfglds1lv-ModelSimSetup-19.1.0.670-linux.run
-r--r--r-- 1 root root       1477 Jan  1  1970 /nix/store/ga6v7kmprbscabh98zwdkl6mvilh8qjw-ModelSimSetup-19.1.0.670-linux.run.drv
-r--r--r-- 1 root root       1477 Jan  1  1970 /nix/store/hj9gvpmzjszrcq5yndq6jjs8n7jrcdkl-ModelSimSetup-19.1.0.670-linux.run.drv
-r--r--r-- 1 root root 1047232539 Jan  1  1970 /nix/store/y75qnagk59qfhpc5sla83iv7g4jnxdb9-ModelSimSetup-19.1.0.670-linux.run
```

**Comment by doronbehar** [comment]
> I think I got vsim working! The overrides are a bit ugly the way they are now though, can this be applied to all of pkgs cleanly somehow?

I don't think it's that bad, but other reviewers may have a stronger say on this then I :)

File: pkgs/applications/editors/quartus-prime/default.nix
Comment (Mic92) [inline pkgs/applications/editors/quartus-prime/default.nix]: I think it is possible to apply overlays with `overrideScope`. I do not have concrete code to do that though.

**Comment by Mic92** [comment]
> > I think I got vsim working! The overrides are a bit ugly the way they are now though, can this be applied to all of pkgs cleanly somehow?
> 
> I don't think it's that bad, but other reviewers may have a stronger say on this then I :)

It is acceptable overlays might add more complexity than this.

**Comment by kwohlfahrt** [comment]
> It's seems as if it's because the file names are different - mine is ./QuartusSetup-19.1.0.670-linux.run and the expected filename is QuartusLiteSetup-19.1.0.670-linux.run.

There is a drop-down menu at the top of the download page, where you can select between "Pro", "Standard" and "Lite" downloads, above the release option. This should be set to "Lite", the "Standard" option results in the "QuartusLiteSetup-${version}.run". This URL should set it appropriately: https://fpgasoftware.intel.com/19.1/?edition=lite&platform=linux

**Comment by Mic92** [comment]
@kwohlfahrt Could you add this as a comment or in the error message?

**Comment by kwohlfahrt** [comment]
Done - I modified the URL to include the edition and version.

**Comment by Ericson2314** [comment]
@expipiplus1 I guess Intel has cleaned it up :)

**Comment by expipiplus1** [comment]
\o/ Thanks, can't wait to try this out.

Anyone reading might also be interested in @bjornfor's quartus packaging: https://github.com/bjornfor/nixos-config/blob/master/pkgs/altera-quartus/default.nix

**Comment by doronbehar** [comment]
@expipiplus1 you can try right away by running `nix-env -f. -iA quartus` inside `nixpkgs/` repo.

**Comment by bjornfor** [comment]
Quartus is in nixpkgs master? Nice!

**Comment by meck** [comment]
@kwohlfahrt would it be possible to add the nios2 eclipse stuff as an optional component? its a manual [install](https://www.intel.com/content/altera-www/global/en_us/index/support/support-resources/knowledge-base/tools/2019/why-does-the-nios--ii-not-installed-after-full-installation-of-t.html). I tried my hand at it but I didn't get far...

**Comment by kwohlfahrt** [comment]
@meck I'm not currently planning to work on the eclipse integration. I haven't used eclispe before, and the instructions linked call for a very old version if eclipse that might be a lot of effort to get working.

**Comment by meck** [comment]
Thats resonable, thanks anyway!

**Pull request merged**
